### PR TITLE
Deprecate DataFrame indexer for iloc setitem and getitem

### LINKED ISSUE #39004: BUG: IndexError: positional indexers are out-of-bounds iloc boolean indexing
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

#### Code Sample, a copy-pastable example

```python
import pandas as pd
df = pd.DataFrame([[0, 1, 2]], columns=['a', 'b', 'c'])
mask = pd.DataFrame([[False, True, False]], columns=['a', 'b', 'c'])
df.iloc[mask] = 3 # Works fine with assignment
print(df.iloc[mask]) # Throws IndexError, but should give similar result as df.values[mask]
```

#### Problem description

Documentation of Pandas says I can use boolean array with iloc. The iloc call with the assignment works fine, but without assignment fails. Just as a note here: In pandas 1.0.1 not even the assignment works (it starts working again if there are more than 1 rows of data).

#### Expected Output
3 (similar as df.values[mask])
#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit           : 3e89b4c4b1580aa890023fc550774e63d499da25
python           : 3.8.5.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.4.0-58-generic
Version          : #64-Ubuntu SMP Wed Dec 9 08:16:25 UTC 2020
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8
pandas           : 1.2.0
numpy            : 1.18.4
pytz             : 2019.3
dateutil         : 2.7.3
pip              : 20.2
setuptools       : 46.2.0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.5.0
html5lib         : 1.0.1
pymysql          : None
psycopg2         : 2.8.5 (dt dec pq3 ext lo64)
jinja2           : 2.11.2
IPython          : 7.19.0
pandas_datareader: None
bs4              : 4.8.2
bottleneck       : None
fsspec           : 0.7.3
fastparquet      : None
gcsfs            : None
matplotlib       : 3.2.1
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyxlsb           : None
s3fs             : None
scipy            : 1.4.1
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : 1.2.0
xlwt             : None
numba            : 0.49.1

</details>

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #39004
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

**Review by jorisvandenbossche — CHANGES_REQUESTED** [review]
See my note on the PR: https://github.com/pandas-dev/pandas/issues/39004#issuecomment-756344692

In addition, if we disallow this, I think it should be deprecated first for assignment (since that case was working right now)

**Comment by jreback** [comment]
> See my note on the PR: [#39004 (comment)](https://github.com/pandas-dev/pandas/issues/39004#issuecomment-756344692)
> 
> In addition, if we disallow this, I think it should be deprecated first for assignment (since that case was working right now)

this is reasonable to deprecate for setitem and simply raise for getitem. @phofl if you can amend.

**Comment by phofl** [comment]
Don't know if I would call this wrong. It is counterintuitive for sure. 

If you *know* that iloc does not align, you may expect the result, Additionally it only works if indexer has the same dimension as the object to index.

Will add a DeprecationWarning

**Comment by jreback** [comment]
> Will add a DeprecationWarning 

FutureWarning but thanks

**Comment by phofl** [comment]
Added the Warning

File: pandas/core/indexing.py [hunk 1]
```diff
@@ -1389,6 +1389,14 @@ def _has_valid_setitem_indexer(self, indexer) -> bool:
         if isinstance(indexer, dict):
             raise IndexError("iloc cannot enlarge its target object")
 
+        if isinstance(indexer, ABCDataFrame):
+            warnings.warn(
+                "DataFrame indexer for iloc is deprecated and will be removed in "
```
Comment (jreback) [inline pandas/core/indexing.py:1394]:
```suggestion
                "DataFrame indexer for .iloc is deprecated and will be removed in
                
                 consider using .loc with a DataFrame indexer for automatic alignment.
                   "
```

Comment (phofl) [reply pandas/core/indexing.py]: Thx

File: pandas/core/indexing.py [hunk 2]
```diff
@@ -1480,6 +1488,9 @@ def _get_list_axis(self, key, axis: int):
             raise IndexError("positional indexers are out-of-bounds") from err
 
     def _getitem_axis(self, key, axis: int):
+        if isinstance(key, ABCDataFrame):
+            raise IndexError("DataFrame indexer is not allowed for iloc")
```
Comment (jreback) [inline pandas/core/indexing.py:1494]:
```suggestion
            raise IndexError("DataFrame indexer is not allowed for .iloc\nConsider using .loc for automatic alignment.")
```

**Comment by phofl** [comment]
Added a deprecation whatsnew

File: pandas/core/indexing.py [hunk 3]
```diff
@@ -1389,6 +1389,16 @@ def _has_valid_setitem_indexer(self, indexer) -> bool:
         if isinstance(indexer, dict):
             raise IndexError("iloc cannot enlarge its target object")
 
+        if isinstance(indexer, ABCDataFrame):
+            warnings.warn(
+                "DataFrame indexer for .iloc is deprecated and will be removed in"
+                "a future version.\n"
+                "consider using .loc with a DataFrame indexer for automatic alignment."
+                "a future version",
```
Comment (jreback) [inline pandas/core/indexing.py:1397]: repeated 'a future version'

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm tiny comment

File: pandas/core/indexing.py (hunk 3 above)
Comment (phofl) [reply pandas/core/indexing.py:1397]: Thx, changed

**Review by jreback — APPROVED** [review]

**Comment by phofl** [comment]
@jreback green

**Comment by jorisvandenbossche** [comment]
Let's give this a bit more time, since the discussion in https://github.com/pandas-dev/pandas/issues/39004 is still ongoing

**Comment by jreback** [comment]
@phofl conflict and if you can rebase. i think this is fine to merge.

**Comment by phofl** [comment]
Done, failure unrelated

**Comment by jreback** [comment]
this looks fine to me. @jorisvandenbossche are you still objecting?

**Comment by jreback** [comment]
@phofl can you merge master and update

**Comment by phofl** [comment]
@jreback merged master, failure unrelated

**Comment by jreback** [comment]
everyone ok here

@jorisvandenbossche @phofl @jbrockmendel @gooney47

**Comment by jbrockmendel** [comment]
LGTM

**Comment by phofl** [comment]
Are we good here?

**Review by jreback — APPROVED** [review]
lgtm. ping on greenish

**Comment by phofl** [comment]
@jreback greenish

**Comment by jreback** [comment]
thanks @phofl

**Pull request merged**